### DISCUSSION TITLE
context-aware lexing clarification

### DISCUSSION THREAD

**Discussion by dfgordon** [tree-sitter/tree-sitter#1596] [Q&A] [answered]
In the [documentation on conflicting tokens](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens), the first disambiguation strategy is "context-aware lexing."  The phrase "valid at that position in the document" is used.  But the position by itself has no context.  Would this be better phrased "valid given the most recently established X" where X is some element of parser state?  If yes, then what is X?  Is it correct to assume there is no lookahead involved in determining which tokens to look for?  Any further info about how it works would be good to know.

**Comment by ahelwer** [comment, marked as answer]
By default tree-sitter is a LR(1) parser, meaning it can do one token of lookahead. So the "X" in this case is the preceding sequence of tokens. That preceding sequence must have only one possible interpretation. If there are multiple possible interpretations, tree-sitter will warn you of this during `tree-sitter generate` and you'll have to either assert static precedence or declare a conflict, which means tree-sitter will use GLR parsing in the conflicting case. GLR parsing splits the parser into multiple copies which each assume a different interpretation and try to parse the text under that interpretation. Usually, only one GLR branch "survives" so that is the one that is chosen in the end. If multiple GLR branches can parse the text without errors then you need to use dynamic precedence to establish which branch is correct, although tree-sitter will not warn you of this possibility during generation.

You can override all of this behavior by using an external scanner, which gives you arbitrary lookahead and the ability to store arbitrary state in the scanner. Usually used for context-sensitive parsing like indentation for a code block in Python.

**Reply by dfgordon** [reply]
Thanks for a good summary.  As a matter of fact I have an external scanner, just tinkering with whether it is truly needed.
About the GLR branching:  Did you mean it gets triggered if and only if `conflicts` is declared?

**Reply by ahelwer** [reply]
Correct.

**Reply by dfgordon** [reply]
I'm not so sure because of the following.  Say I have a language where `if` is a keyword and also allowed to be a variable.  The following is valid in this toy language:
```
if=if
ifif
```
I was expecting that LR(1) would eat the first `if` and be forced to decide then and there whether it is a variable or keyword.  But TS can parse this fine without any `conflicts` declaration:
```js
module.exports = grammar({
	name: 'tiny_if',
	extras: $ => [' '],
	rules: {
		source_file: $ => repeat($.line),
		line: $ => seq(choice($.assignment,$.conditional),$.newline),
		assignment: $ => seq($.var,'=',$.var),
		conditional: $ => seq($.if,$.var),
		var: $ => choice('if',repeat1(/[a-z]/)),
		if: $ => 'if',
		newline: $ => /\r?\n/
	}
});
```
Output of `tree-sitter parse` looks perfect:
```
(source_file [0, 0] - [2, 0]
  (line [0, 0] - [1, 0]
    (assignment [0, 0] - [0, 5]
      (var [0, 0] - [0, 2])
      (var [0, 3] - [0, 5]))
    (newline [0, 5] - [1, 0]))
  (line [1, 0] - [2, 0]
    (conditional [1, 0] - [1, 4]
      (if [1, 0] - [1, 2])
      (var [1, 2] - [1, 4]))
    (newline [1, 4] - [2, 0])))
```

**Reply by ahelwer** [reply]
I think my explanation was a bit imprecise about what I meant by interpretation. In my mind it's really like the grammar is compiled to a state machine (which indeed it is) and the grammar can enter the "has consumed if" state without needing to decide what it means. At that point there's a branch: either it can consume a "=", or it can consume a variable name. So far so good, since it can decide that branch purely by looking at the next token. And this is where I get a bit fuzzy, but I think if you needed to consume *two* tokens before coming to this branch instead of just one, that would be a conflict. Or something. I think it's also relevant whether you are identifying terminal or intermediate rule instances. Maybe someone else can explain!

**Reply by dfgordon** [reply]
Confirmed, using `parse --debug` shows that `version_count` never changes from 1 (so presumably 1 branch only), and it shows `reduce sym:var` happening right after the `=` gets consumed.  I think I just misunderstood LR(1).

>  it's also relevant whether you are identifying terminal or intermediate rule

Confirmed, if `repeat1(/[a-z]/)` is replaced by `/[a-z]+/` it fails to parse.

**Reply by dfgordon** [reply]
Note for the record: after more experiments I'm not sure if `version_count` is an indicator of GLR branching.  In particular, I've seen it take values > 1 without `conflicts`.

**Comment by stephe-ada-guru** [comment]
dfgordon ***@***.***> writes:

> I'm not so sure because of the following.  Say I have a language where `if` is a keyword and also allowed to be a variable.  The following is valid in this toy language:
> ```
> if=if
> ifif
> ```
> I was expecting that LR(1) would eat the first `if` and be forced to
> decide then and there whether it is a variable or keyword.  But TS can
> parse this fine without any `conflicts` declaration:

The _grammar_ has no notion of "keyword" vs "variable"; it just has
tokens and consists of rules about where they can occur.

We like to think of "keyword" and "variable" because the _compiler_
makes those distinctions at a higher level.

-- 
-- Stephe

**Reply by dfgordon** [reply]
Agreed, I should have said it has to decide between the two grammar rules `var` and `if`.
